Replace ResourceResponse with ResourceResponse::CrossThreadData in CacheStorage messages

szewai opened the pull request:
#### 2f169036ec502e5451e1a00f5d9d268cf4ffe6bd
<pre>
Replace ResourceResponse with ResourceResponse::CrossThreadData in CacheStorage messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=254231">https://bugs.webkit.org/show_bug.cgi?id=254231</a>
rdar://107020237

Reviewed by Youenn Fablet.

NetworkStorageManager receives and handles CacheStorage messages on a WorkQueue, which is not tied to one thread.
ResourceResponse contains AtomSring members, which can only be accessed from the thread where they are created. To avoid
threading issues, we replace ResourceResponse with ResourceResponse::CrossThreadData in CacheStorage messages that are
sent to and sent from NetworkStorageManager.

To make this change, this patch introduces DOMCacheEngine::CrossThreadRecord and adds encode/decode functions for
ResourceResponse::CrossThreadData. It also adds members to ResourceResponse::CrossThreadData so that it contains all
information needed for persistence encoding/decoding.

* Source/WebCore/Modules/cache/DOMCacheEngine.cpp:
(WebCore::DOMCacheEngine::toCrossThreadRecord):
(WebCore::DOMCacheEngine::fromCrossThreadRecord):
* Source/WebCore/Modules/cache/DOMCacheEngine.h:
(WebCore::DOMCacheEngine::CrossThreadRecord::CrossThreadRecord):
* Source/WebCore/loader/CrossOriginAccessControl.cpp:
(WebCore::shouldCrossOriginResourcePolicyCancelLoad):
(WebCore::validateCrossOriginResourcePolicy):
* Source/WebCore/loader/CrossOriginAccessControl.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::CrossThreadData::isolatedCopy const):
(WebCore::ResourceResponseBase::crossThreadData const):
(WebCore::ResourceResponseBase::fromCrossThreadData):
(WTF::Persistence::Coder&lt;WebCore::ResourceResponseBase::CrossThreadData&gt;::encode):
(WTF::Persistence::Coder&lt;WebCore::ResourceResponseBase::CrossThreadData&gt;::decode):
* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseBase::CrossThreadData::CrossThreadData):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
(WebKit::toCacheStorageRecord):
(WebKit::CacheStorageCache::retrieveRecords):
(WebKit::CacheStorageCache::putRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
(WebKit::decodeRecordHeader):
(WebKit::readRecordInfoFromFileData):
(WebKit::CacheStorageDiskStore::readRecordFromFileData):
(WebKit::encodeRecordHeader):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::cacheStorageRetrieveRecords):
(WebKit::NetworkStorageManager::cacheStoragePutRecords):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(generate_headers_for_header):
(generate_header):
(parse_serialized_types):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchPutOperation):

Canonical link: <a href="https://commits.webkit.org/262102@main">https://commits.webkit.org/262102@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99a4baab41944107cd89a9bf0da4d69e0fa90fd7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/471 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/532 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/640 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/718 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/518 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/520 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/624 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/558 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/506 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/544 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/476 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/527 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/507 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/522 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/132 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/516 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 51;
```
not needed

youennf commented on Source/WebCore/loader/CrossOriginAccessControl.cpp:
```diff
@@ -330,16 +329,28 @@ static inline bool shouldCrossOriginResourcePolicyCancelLoad(CrossOriginEmbedder
+    return shouldCrossOriginResourcePolicyCancelLoad(coep, origin, response.isNull(), response.url(), response.httpHeaderField(HTTPHeaderName::CrossOriginResourcePolicy), forNavigation);
+}
+
+std::optional<ResourceError> validateCrossOriginResourcePolicy(CrossOriginEmbedderPolicyValue coep, const SecurityOrigin& origin, const URL& requestURL, const ResourceResponse::CrossThreadData& responseData, ForNavigation forNavigation)
```
I would be tempted to pass response url and response CORP header instead of cross thread data

youennf commented on Source/WebCore/loader/CrossOriginAccessControl.cpp:
```diff
@@ -330,16 +329,28 @@ static inline bool shouldCrossOriginResourcePolicyCancelLoad(CrossOriginEmbedder
+    return std::nullopt;
+}
+
 std::optional<ResourceError> validateCrossOriginResourcePolicy(CrossOriginEmbedderPolicyValue coep, const SecurityOrigin& origin, const URL& requestURL, const ResourceResponse& response, ForNavigation forNavigation)
```
If we add a variant, let's make one variant call the other.

youennf commented on Source/WebCore/Modules/cache/DOMCacheEngine.h:
```diff
@@ -80,6 +80,26 @@ struct Record {
 };
 
+struct CrossThreadRecord {
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
```
We should make it non copyable.

youennf commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -469,6 +497,129 @@ bool ResourceResponseBase::decode(Decoder& decoder, ResourceResponseBase& respon
     return true;
 }
 
+template<class Encoder, typename> void ResourceResponseBase::CrossThreadData::encode(Encoder& encoder) const
```
This is used for storage, so maybe update the names accordingly.
Maybe mention that changing this should change the cache storage version number.

szewai replied:
I have moved this to WTF::Persistence, so it's more clear that it is used for storage and not supposed to be updated without checking disk storage.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -243,8 +244,13 @@ static std::optional<RecordHeader> decodeRecordHeader(Span<const uint8_t> header
+    if (!isNull || *isNull)
+        return std::nullopt;
+
+    auto responseData = ResourceResponseBase::CrossThreadData::decode(decoder);
```
Should we bump the version number?

szewai replied:
I think we don't need to as the disk storage format does not change.

youennf replied:
OK, please make sure this is actually the case by trying to load past stored data with the new decoder.
I guess this might explain the isNull encoding?
If so, maybe add a FIXME that clearly says to remove isNull when we bump the version number

szewai replied:
Yes, isNull is encoded and decoded to keep the same disk storage format.
I've manually tested and verified the old CacheStorage file could be decoded correctly.

youennf commented on Source/WebKit/Shared/WebCoreArgumentCoders.h:
```diff
@@ -225,6 +226,11 @@ template<> struct ArgumentCoder<WebCore::DOMCacheEngine::Record> {
+template<> struct ArgumentCoder<WebCore::DOMCacheEngine::CrossThreadRecord> {
+    static void encode(Encoder&, const WebCore::DOMCacheEngine::CrossThreadRecord&);
+    static std::optional<WebCore::DOMCacheEngine::CrossThreadRecord> decode(Decoder&);
+};
```
Why cannot we use WebCoreArgumentCoders.serialization.in?

youennf commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
```diff
@@ -77,7 +77,16 @@ void WebCacheStorageConnection::retrieveCaches(const WebCore::ClientOrigin& orig
 void WebCacheStorageConnection::retrieveRecords(WebCore::DOMCacheIdentifier cacheIdentifier, WebCore::RetrieveRecordsOptions&& options, WebCore::DOMCacheEngine::RecordsCallback&& callback)
 {
-    connection().sendWithAsyncReply(Messages::NetworkStorageManager::CacheStorageRetrieveRecords(cacheIdentifier, options), WTFMove(callback));
+    auto newCallback = [callback = WTFMove(callback)](auto&& result) mutable {
```
I am not sure we want this.
Since we have cross thread records, we could optimise worker connection code by doing a simple move or at least an optimised isolatedCopy()&&?

szewai replied:
This patch aims to fix network process crashes, so I hope to limit the amount of changes to reduce risk. I will make a followup for updating code on web process side.

youennf commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
```diff
@@ -87,7 +96,10 @@ void WebCacheStorageConnection::batchDeleteOperation(WebCore::DOMCacheIdentifier
 {
-    connection().sendWithAsyncReply(Messages::NetworkStorageManager::CacheStoragePutRecords(cacheIdentifier, records), WTFMove(callback));
+    auto crossThreadRecords = WTF::map(records, [](auto&& record) {
+        return toCrossThreadRecord(WTFMove(record));
```
We could also update this code to directly pass cross thread records to this method.

szewai replied:
I could make a followup to update the client-side code.

youennf reviewed (approved):
Please also file clean-up bugs so that we keep track of them

szewai committed:
Replace ResourceResponse with ResourceResponse::CrossThreadData in CacheStorage messages
https://bugs.webkit.org/show_bug.cgi?id=254231
rdar://107020237

Reviewed by Youenn Fablet.

NetworkStorageManager receives and handles CacheStorage messages on a WorkQueue, which is not tied to one thread.
ResourceResponse contains AtomSring members, which can only be accessed from the thread where they are created. To avoid
threading issues, we replace ResourceResponse with ResourceResponse::CrossThreadData in CacheStorage messages that are
sent to and sent from NetworkStorageManager.

To make this change, this patch introduces DOMCacheEngine::CrossThreadRecord and adds encode/decode functions for
ResourceResponse::CrossThreadData. It also adds members to ResourceResponse::CrossThreadData so that it contains all
information needed for persistence encoding/decoding.

* Source/WebCore/Modules/cache/DOMCacheEngine.cpp:
(WebCore::DOMCacheEngine::toCrossThreadRecord):
(WebCore::DOMCacheEngine::fromCrossThreadRecord):
* Source/WebCore/Modules/cache/DOMCacheEngine.h:
(WebCore::DOMCacheEngine::CrossThreadRecord::CrossThreadRecord):
* Source/WebCore/loader/CrossOriginAccessControl.cpp:
(WebCore::shouldCrossOriginResourcePolicyCancelLoad):
(WebCore::validateCrossOriginResourcePolicy):
* Source/WebCore/loader/CrossOriginAccessControl.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::CrossThreadData::isolatedCopy const):
(WebCore::ResourceResponseBase::crossThreadData const):
(WebCore::ResourceResponseBase::fromCrossThreadData):
(WTF::Persistence::Coder<WebCore::ResourceResponseBase::CrossThreadData>::encode):
(WTF::Persistence::Coder<WebCore::ResourceResponseBase::CrossThreadData>::decode):
* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseBase::CrossThreadData::CrossThreadData):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
(WebKit::toCacheStorageRecord):
(WebKit::CacheStorageCache::retrieveRecords):
(WebKit::CacheStorageCache::putRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
(WebKit::decodeRecordHeader):
(WebKit::readRecordInfoFromFileData):
(WebKit::CacheStorageDiskStore::readRecordFromFileData):
(WebKit::encodeRecordHeader):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::cacheStorageRetrieveRecords):
(WebKit::NetworkStorageManager::cacheStoragePutRecords):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(generate_headers_for_header):
(generate_header):
(parse_serialized_types):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchPutOperation):

Canonical link: https://commits.webkit.org/262102@main
